core/nandtool/flashmng.c

```diff
@@ -0,0 +1,171 @@
+/*
+ * Phoenix-RTOS
+ *
+ * NANDtool utility
+ *
+ * Flash Server Manager
+ *
+ * Copyright 2019, 2021 Phoenix Systems
+ * Author: Hubert Buczynski, Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdio.h>
+
+#include <sys/file.h>
+#include <sys/msg.h>
+
+#include "flashmng.h"
+
+
+static struct {
+	oid_t oid;
+	flashsrv_info_t info;
+} flashmng_common;
+
+
+int flashmng_writeraw(oid_t oid, unsigned int page, void *data, size_t size)
+{
+	msg_t msg = { 0 };
+	flash_i_devctl_t *idevctl = (flash_i_devctl_t *)msg.i.raw;
+	volatile flash_o_devctl_t *odevctl = (flash_o_devctl_t *)msg.o.raw;
+	flashsrv_info_t *info;
+	int err;
+
+	if ((info = flashmng_info(oid)) == NULL)
+		return -EFAULT;
+
+	msg.type = mtDevCtl;
+	msg.i.data = data;
+	msg.i.size = size;
+	idevctl->type = flashsrv_devctl_writeraw;
+	idevctl->write.oid = oid;
+	idevctl->write.address = page * (info->metasz + info->writesz);
+	idevctl->write.size = size;
+
+	if (((err = msgSend(oid.port, &msg)) < 0) || ((err = odevctl->err) < 0))
+		return err;
+
+	if (err != size)
+		return -EIO;
+
+	return EOK;
+}
+
+
+int flashmng_erase(oid_t oid, unsigned int start, unsigned int end)
+{
+	msg_t msg = { 0 };
+	flash_i_devctl_t *idevctl = (flash_i_devctl_t *)msg.i.raw;
+	volatile flash_o_devctl_t *odevctl = (flash_o_devctl_t *)msg.o.raw;
+	flashsrv_info_t *info;
+	int err;
+
+	if ((info = flashmng_info(oid)) == NULL)
+		return -EFAULT;
+
+	msg.type = mtDevCtl;
+	idevctl->type = flashsrv_devctl_erase;
+	idevctl->erase.oid = oid;
+	idevctl->erase.address = start * info->erasesz;
+	idevctl->erase.size = (end - start) * info->erasesz;
+
+	if (((err = msgSend(oid.port, &msg)) < 0) || ((err = odevctl->err) < 0))
+		return err;
+
+	return EOK;
+}
+
+
+int flashmng_checkbad(oid_t oid)
+{
+	flashsrv_info_t *info;
+	unsigned int block, nbad = 0;
+	int err;
+
+	if ((info = flashmng_info(oid)) == NULL)
+		return -EFAULT;
+
+	for (block = 0; block < info->size / info->erasesz; block++) {
+		if ((err = flashmng_isbad(oid, block))) {
+			if (err < 0)
+				return err;
+
+			printf("Block %u is marked as bad\n", block);
+			nbad++;
+		}
+	}
+
+	printf("Total blocks checked: %u\n", block);
+	printf("Number of bad blocks:  %u\n", nbad);
+
+	if (nbad >= 256)
```

Comment:
this is not true now (previously `DBBT` could hold only 256 records per our implementation)

Reply:
Max `DBBT` size depends on block size of a given flash device (we typically use one block for `DBBT`). I'm leaving it to be fixed in another PR - we currently don't even use `DBBT`.